### ISSUE TITLE
FM7: help file doesn't specify "algo" must be fixed

### ISSUE THREAD

**Issue by Arseniusz** [supercollider/sc3-plugins#151]
```SuperCollider
    SynthDef.new(\fm_synth, {
        |freq, alg = 0|
        var sig, freq_matrix, envelope;
        freq_matrix = [
            // freq, phase, amp
            [freq, 0, 1],
            [freq * 2, 0, 0.5],
            [freq * 4, 0, 0.25],
            [5009, 0, 0.1],
            [20, 0, 1],
            [30, 0, 1]
        ];
        envelope = EnvGen.kr(Env.perc(level: 0.03), doneAction: 2);
        sig = FM7.arAlgo(alg, freq_matrix, 1.5) * envelope;
        Out.ar(0, sig);
    }).add;
```

If I trying to send 'algo' argument in FM7.arAlgo I got an error:

ERROR: Primitive '_BasicAt' failed.
Index not an Integer
RECEIVER:
Instance of Array {    (0x55f9a39fdb38, gc=F4, fmt=01, flg=00, set=05)
  indexed slots [32]
      0 : instance of Function (0x55f9a424e258, size=2, set=2)
      1 : instance of Function (0x55f9a4616938, size=2, set=2)
      2 : instance of Function (0x7fbe280813f8, size=2, set=2)
      3 : instance of Function (0x55f9a4a82a78, size=2, set=2)
      4 : instance of Function (0x55f9a4619aa8, size=2, set=2)
      5 : instance of Function (0x55f9a49dbea8, size=2, set=2)
      6 : instance of Function (0x55f9a34bccc8, size=2, set=2)
      7 : instance of Function (0x55f9a4656808, size=2, set=2)
      8 : instance of Function (0x55f9a44168a8, size=2, set=2)
      9 : instance of Function (0x55f9a4a0acb8, size=2, set=2)
     10 : instance of Function (0x55f9a4a0e988, size=2, set=2)
     11 : instance of Function (0x55f9a4a16ef8, size=2, set=2)
     12 : instance of Function (0x55f9a48c33a8, size=2, set=2)
     13 : instance of Function (0x55f9a49029a8, size=2, set=2)
     14 : instance of Function (0x55f9a468fcf8, size=2, set=2)
     15 : instance of Function (0x55f9a4a05378, size=2, set=2)
     16 : instance of Function (0x55f9a46500c8, size=2, set=2)
     17 : instance of Function (0x55f9a468d678, size=2, set=2)
     18 : instance of Function (0x7fbe28046088, size=2, set=2)
     19 : instance of Function (0x55f9a46163f8, size=2, set=2)
     20 : instance of Function (0x7fbe2000b648, size=2, set=2)
     21 : instance of Function (0x55f9a4a88568, size=2, set=2)
     22 : instance of Function (0x55f9a4b489a8, size=2, set=2)
     23 : instance of Function (0x55f9a4b5dcd8, size=2, set=2)
     24 : instance of Function (0x55f9a3f5c088, size=2, set=2)
     25 : instance of Function (0x7fbe28043fc8, size=2, set=2)
     26 : instance of Function (0x55f9a4aa0058, size=2, set=2)
     27 : instance of Function (0x55f9a468e6a8, size=2, set=2)
     28 : instance of Function (0x55f9a4592918, size=2, set=2)
     29 : instance of Function (0x7fbe28099908, size=2, set=2)
     30 : instance of Function (0x7fbe2000db28, size=2, set=2)
     31 : instance of Function (0x7fbe2001f458, size=2, set=2)
}
PATH: /home/dude/Documents/supercollider/no_needs_to_go.scd

PROTECTED CALL STACK:
	Meta_MethodError:new	0x55f99fc07b00
		arg this = PrimitiveFailedError
		arg what = Index not an Integer
		arg receiver = [ a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function ]
	Meta_PrimitiveFailedError:new	0x55f99fc0e040
		arg this = PrimitiveFailedError
		arg receiver = [ a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function ]
	Object:primitiveFailed	0x55f9a13f66c0
		arg this = [ a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function ]
	Meta_FM7:algoSpec	0x55f9a162e500
		arg this = FM7
		arg algo = an OutputProxy
		arg feedback = 1.5
		var matrix = [ [ 0, 0, 0, 0, 0, 0 ], [ 0, 0, 0, 0, 0, 0 ], [ 0, 0, 0, 0, 0, 0 ], [ 0, 0, 0, 0, 0, 0 ], [ 0, 0, 0, 0, 0, 0 ], [ 0, 0, 0, 0, 0, 0 ] ]
	Meta_FM7:arAlgo	0x55f9a1635c40
		arg this = FM7
		arg algo = an OutputProxy
		arg ctlMatrix = [ [ an OutputProxy, 0, 1 ], [ a BinaryOpUGen, 0, 0.5 ], [ a BinaryOpUGen, 0, 0.25 ], [ 5009, 0, 0.1 ], [ 20, 0, 1 ], [ 30, 0, 1 ] ]
		arg feedback = 1.5
		var modMatrix = nil
		var channels = nil
	a FunctionDef	0x55f99fecd1e8
		sourceCode = "{
        |freq, algo|
        var sig, freq_matrix, envelope;
        freq_matrix = [
            // freq, phase, amp
            [freq, 0, 1],
            [freq * 2, 0, 0.5],
            [freq * 4, 0, 0.25],
            [5009, 0, 0.1],
            [20, 0, 1],
            [30, 0, 1]
        ];
        envelope = EnvGen.kr(Env.perc(level: 0.03), doneAction: 2);
        sig = FM7.arAlgo(algo, freq_matrix, 1.5) * envelope;
        Out.ar(0, sig);
    }"
		arg freq = an OutputProxy
		arg algo = an OutputProxy
		var sig = nil
		var freq_matrix = [ [ an OutputProxy, 0, 1 ], [ a BinaryOpUGen, 0, 0.5 ], [ a BinaryOpUGen, 0, 0.25 ], [ 5009, 0, 0.1 ], [ 20, 0, 1 ], [ 30, 0, 1 ] ]
		var envelope = an EnvGen
	SynthDef:buildUgenGraph	0x55f9a2db9180
		arg this = a SynthDef
		arg func = a Function
		arg rates = nil
		arg prependArgs = [  ]
		var result = nil
		var saveControlNames = nil
	a FunctionDef	0x55f9a2db77c0
		sourceCode = "<an open Function>"
	Function:prTry	0x55f9a2274100
		arg this = a Function
		var result = nil
		var thread = a Thread
		var next = nil
		var wasInProtectedFunc = false
	
CALL STACK:
	MethodError:reportError   0x55f9a3b90e58
		arg this = <instance of PrimitiveFailedError>
	Nil:handleError   0x55f9a3b85a28
		arg this = nil
		arg error = <instance of PrimitiveFailedError>
	Thread:handleError   0x55f9a471bed8
		arg this = <instance of Thread>
		arg error = <instance of PrimitiveFailedError>
	Object:throw   0x55f9a3b99b68
		arg this = <instance of PrimitiveFailedError>
	Function:protect   0x55f9a3f966d8
		arg this = <instance of Function>
		arg handler = <instance of Function>
		var result = <instance of PrimitiveFailedError>
	SynthDef:build   0x55f9a3554698
		arg this = <instance of SynthDef>
		arg ugenGraphFunc = <instance of Function>
		arg rates = nil
		arg prependArgs = nil
	< closed FunctionDef >  (no arguments or variables)
	Interpreter:interpretPrintCmdLine   0x7fbe280e0288
		arg this = <instance of Interpreter>
		var res = nil
		var func = <instance of Function>
		var code = "    SynthDef.new(\piano, {
 ..."
		var doc = nil
		var ideClass = <instance of Meta_ScIDE>
	Process:interpretPrintCmdLine   0x55f9a4780ba8
		arg this = <instance of Main>
^^ The preceding error dump is for ERROR: Primitive '_BasicAt' failed.
Index not an Integer
RECEIVER: [ a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function, a Function ]

**Comment by shimpe** [comment]
I'm not a specialist, but it seems the problem is in the surrounding code you are not showing? From the stack trace I think I understand that algo must be an integer and you are passing in something else (an OutputProxy).

**Comment by Arseniusz** [comment]
@shimpe of course alg = 0, fix in issue. Same problem.
```Supercollider
    SynthDef.new(\fm_synth, {
        |freq, alg = 0|
        var sig, freq_matrix, envelope;
        freq_matrix = [
            // freq, phase, amp
            [freq, 0, 1],
            [freq * 2, 0, 0.5],
            [freq * 4, 0, 0.25],
            [5009, 0, 0.1],
            [20, 0, 1],
            [30, 0, 1]
        ];
        envelope = EnvGen.kr(Env.perc(level: 0.03), doneAction: 2);
        sig = FM7.arAlgo(alg, freq_matrix, 1.5) * envelope;
        Out.ar(0, sig);
    }).add;
```

**Comment by shimpe** [comment]
Upon reconsideration, the problem is probably that changing algo you try to do would require changing the UGen graph on the server dynamically and that is not supported in supercollider.

However, there usually are alternative ways of tackling the problem. One such alternative could be to create a SynthDef for every possible value of algo, and instantiate the correct Synths based on the value of alg. 

Here's an example of what I mean. Not sure if it will help you accomplish your own goals though...

Basically a new SynthDef is created for all 32 values of algo (each with unique names:
\fm_synth0, \fm_synth1, ..., \fm_synth31 ) and then later on the correct synth is instantiated
based on the value of "value_of_alg". 

```
(
~makeSynth = { | alg = 0 |	
	SynthDef.new(("fm_synth" ++ alg).asSymbol, {
		|freq = 440 |
		var sig, freq_matrix, envelope;
		freq_matrix = [
			// freq, phase, amp
			[freq, 0, 1],
			[freq * 2, 0, 0.5],
			[freq * 4, 0, 0.25],
			[5009, 0, 0.1],
			[20, 0, 1],
			[30, 0, 1]
		];
		
		envelope = EnvGen.kr(Env.perc(level: 0.03), doneAction: 2);
		sig = FM7.arAlgo(alg, freq_matrix, 1.5) * envelope;
		Out.ar(0, sig);
	}).add;	
};
~synths = Array.newClear(32);
32.do({
	| index |
	~makeSynth.value(index);
});
fork {
	~value_of_alg = 3;
	Synth.new(("fm_synth"++~value_of_alg).asSymbol, [\freq, 440]);
	2.wait;
	~value_of_alg = 15;
	Synth.new(("fm_synth"++~value_of_alg).asSymbol, [\freq, 660]);
	2.wait;
};
)
```

**Comment by shimpe** [comment]
Instead of instantiating all the synths manually as I did in the above "fork" section, you probably ought to look into supercollider "patterns" which would make it easy to calculate values for "value_of_alg" and use them to instantiate the correct synths on the fly.

Have fun!

**Comment by jamshark70** [comment]
> Upon reconsideration, the problem is probably that changing algo the way you try to it do would require changing the UGen graph on the server dynamically and that is not supported in supercollider.

This is exactly the problem. The FM algorithm defines a number of UGens and their connections. The UGens and connections must be hard coded into the SynthDef. There is no exception to this rule (but users are always looking for exceptions to it).

Therefore the algorithm number must be chosen when building the SynthDef. It is valid to produce a separate SynthDef for each algorithm, as shown above.

So, the initial request as written is impossible (without totally rewriting the UGen to do all the work in C++, but I wouldn't hold my breath for that).

I'll keep this open because the FM7 help file does not document that `algo` must be a fixed value.

**Comment by LFSaw** [comment]
closed with #152 !?
